### ISSUE TITLE
Suggestion: Guidelines on contributing tests

### ISSUE THREAD

**Issue by RamblingCookieMonster** [PowerShell/DscResources#9]
Hi all!

Wondering if it would be possible to provide some guidelines around contributing tests for the DSC Resources.  You probably have a better idea of what would go in this, but a few example questions that might be helpful to answer:

* Do you want integration tests? (e.g. the AppVeyor VM is configured with the resource and the desired state verified via Pester)
* If you want integration tests, I'm assuming you might want a separate file and naming convention?
* How do you want the tests themselves organized in general (e.g. describe/context/it)?

An initial complete test suite for a single repository might go a long way.

Cheers!

**Comment by vors** [comment]
Agree. We have it in plans.

**Comment by tysonjhayes** [comment]
+1 to this. I'd love to contribute some tests to the system and would be very interested in seeing Integration tests specifically.

**Comment by vors** [comment]
I would like to hear some opinions from people with practical knowledge (I invite @BrianFarnhill @TravisEz13 join this thread).

To address @RamblingCookieMonster original questions:

* Integration tests — absolutely. Would be wonderful, if you can provide them, unfortunately, it may be too expensive in some cases (may require cross-machine sync, yada yada).
* It make sense to put integration tests in a separate file.
* I would suggest this layout: 
    - One folder `Tests` in the root of the repo. 
    - If integration tests are presented, I would create sub-folders `Tests\UnitTests` and `Tests\IntegrationTests`. 
    - For every resource `xMyAwesomeResource` in the module, I would create a file `Tests\UnitTests\xMyAwesomeResource.Tests.ps1` with Unit Tests and `Tests\IntegrationTests\xMyAwesomeResource.Integration.Tests.ps1`.
   - Then every `.ps1` file can look as following:

```powershell
Describe 'xMyAwesomeResource' {
  Context 'when there is no awesome feature installed' {
    # Mock all the things, if it's unit tests
    
    # Cleanup logic, to make sure there is no artifacts from previous runs
    It 'install awesome feature on the fresh box' {
      # ...
    }
    # ...  
  }
  
  Context 'when it need to modify awesome feature parameters' {
    # ...
  }
}
```

@BrianFarnhill did some tests in his module https://github.com/PowerShell/xSharePoint
We may try to make it an example for other repos.

**Comment by BrianFarnhill** [comment]
Yea there is no way we can go near the integration tests with the xSharePoint resource easily - a typical run on even a single server SharePoint deploy takes around an hour to an hour and a half to pull off. So we might look at writing them because I do love the idea, but if we do there is no way we can pull that in to AppVeyor (and thats before we hit a small farm deployment which needs multiple servers and goes for 3-4 hours). 

There are also some other issues I have with unit testing the SharePoint ones as well, in that all of the SharePoint powershell cmdlets are all provided through a snapin, not a module, and the only way to get them on a machine is to install SharePoint - which means I can't mock the SharePoint cmdlets to include in my tests. So while the xSharePoint module has a lot of tests we are really looking more at the test methods on each DSC resource and ignoring the setters. There are some other tests I need to include in whats there, but very keen to expand to having some integration tests that we can run manually at least so we have a consistent mechanism for verifying things.

**Comment by TravisEz13** [comment]
@BrianFarnhill ping me offline about a workaround for mocking.

I think we should focus on good unit tests automation before focusing on integration test automation.  
Also, many scenarios like Brian's will require unit testing/mocking to test effectively.  In other words, we should have unit test with good coverage that fail fast, and a few key integration tests that cover important cases.

That being said, some things are easier to cover, or can only be found through integration testing and we should support it.

On the directory structure, I think we should not duplicate the word test where we don't have to.  
* Therefore, I would suggest this layout: 
    - One folder `Tests` in the root of the repo. 
    - If integration tests are presented, I would create sub-folders `Tests\Unit` and `Tests\Integration`. 
    - For every resource `xMyAwesomeResource` in the module, I would create a file `Tests\Unit\xMyAwesomeResource.Tests.ps1` with Unit Tests and `Tests\Integration\xMyAwesomeResource.Integration.Tests.ps1`.

I agree with the separate directory for unit and integration tests.  It'll make it very easy to run just one type of test.

**Comment by BrianFarnhill** [comment]
I agree with the separation of folders as @TravisEz13 suggested - I'll line the xSharePoint resource up with this as part of an upcoming change I need to make to some other tests.

**Comment by TravisEz13** [comment]
We should also document that the tests run elevated by default

**Comment by KarolKaczmarek** [comment]
Closing as this seems to be resolved by #24
